File: bridge/lora_bridge.js
/**
 * =============================================================================
 * SIH Disaster-Response Drone System - Person 2 Ground Station
 * Physical LoRa Receiver Serial Bridge Script (Node.js)
 * =============================================================================
 * Usage:
 *   node bridge/lora_bridge.js
 *
 * Environment Variables:
 *   SERIAL_PORT : Serial port name (Default: 'COM3' or '/dev/ttyUSB0')
 *   BAUD_RATE   : Baud rate (Default: 9600)
 *   BACKEND_URL : Backend endpoint (Default: 'http://localhost:5000/api/events/lora')
 */

const http = require('http');
const os = require('os');

const SERIAL_PORT_NAME = process.env.SERIAL_PORT || (os.platform() === 'win32' ? 'COM3' : '/dev/ttyUSB0');
const BAUD_RATE = parseInt(process.env.BAUD_RATE || '9600', 10);
const BACKEND_URL = process.env.BACKEND_URL || 'http://localhost:5000/api/events/lora';
const IS_SIMULATION = process.argv.includes('--simulate') || process.env.SIMULATE === 'true';

let SerialPort = null;
let ReadlineParser = null;

try {
  const serialport = require('serialport');
  SerialPort = serialport.SerialPort;
  ReadlineParser = require('@serialport/parser-readline').ReadlineParser;
} catch (e) {
  // Optional dependency check
}

function processLine(lineStr) {
  const lineClean = lineStr.trim();
  if (!lineClean) return;

  console.log(`\n[PACKET RECEIVED] Raw input: ${lineClean}`);

  let packet;
  try {
    packet = JSON.parse(lineClean);
    console.log(`[PACKET PARSED] Event ID: ${packet.event_id || 'AUTO'} | Hazard: ${packet.hazard} | Lat/Lng: (${packet.latitude}, ${packet.longitude})`);
  } catch (err) {
    console.error(`❌ [MALFORMED PACKET] Failed to parse JSON: ${err.message}`);
    return;
  }

  // Validate required fields
  if (!packet.hazard || packet.latitude === undefined || packet.longitude === undefined) {
    console.warn(`⚠️ [INVALID PAYLOAD] Missing required fields 'hazard', 'latitude', or 'longitude'`);
    return;
  }

  // Forward payload to Express Backend
  const payloadStr = JSON.stringify(packet);
  const parsedUrl = new URL(BACKEND_URL);

  const reqOptions = {
    hostname: parsedUrl.hostname,
    port: parsedUrl.port || 80,
    path: parsedUrl.pathname,
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': Buffer.byteLength(payloadStr)
    }
  };

  const req = http.request(reqOptions, (res) => {
    let body = '';
    res.on('data', chunk => body += chunk);
    res.on('end', () => {
      if (res.statusCode >= 200 && res.statusCode < 300) {
        console.log(`✅ [PACKET FORWARDED] Backend HTTP ${res.statusCode}: ${body}`);
      } else {
        console.error(`❌ [BACKEND ERROR] Backend HTTP ${res.statusCode}: ${body}`);
      }
    });
  });

  req.on('error', (err) => {
    console.error(`❌ [BACKEND ERROR] Failed to reach backend endpoint (${BACKEND_URL}): ${err.message}`);
  });

  req.write(payloadStr);
  req.end();
}

function main() {
  console.log('=============================================================================');
  printHeader();
  console.log('=============================================================================\n');

  if (IS_SIMULATION) {
    console.log('💡 [SIMULATION MODE ACTIVE] STDIN line listener connected. Type or pipe JSON lines below:\n');
    process.stdin.setEncoding('utf8');
    process.stdin.on('data', (data) => {
      const lines = data.split('\n');
      lines.forEach(processLine);
    });
    return;
  }

  if (!SerialPort) {
    console.log('⚠️ [RECEIVER STATUS] Node.js "serialport" module is not installed.');
    console.log('   👉 For physical USB/UART LoRa hardware, install: npm install serialport');
    console.log('   👉 For simulation mode, run: npm run bridge:lora:sim\n');
    return;
  }

  try {
    console.log(`🔌 Attempting serial connection to [${SERIAL_PORT_NAME}] at ${BAUD_RATE} baud...`);
    const port = new SerialPort({ path: SERIAL_PORT_NAME, baudRate: BAUD_RATE });
    const parser = port.pipe(new ReadlineParser({ delimiter: '\n' }));

    port.on('open', () => {
      console.log(`✅ [RECEIVER CONNECTED] Successfully listening on ${SERIAL_PORT_NAME}`);
      console.log('   Ready for incoming LoRa packet frames from Person 1 drone...\n');
    });

    parser.on('data', processLine);

    port.on('close', () => {
      console.log(`❌ [RECEIVER DISCONNECTED] Serial port closed.`);
    });

    port.on('error', (err) => {
      console.error(`❌ [RECEIVER ERROR] Serial error on ${SERIAL_PORT_NAME}: ${err.message}`);
    });

  } catch (err) {
    console.error(`❌ [RECEIVER INITIALIZATION ERROR] ${err.message}`);
  }
}

function printHeader() {
  if (IS_SIMULATION) {
    console.log('📡 SIH DRONE GROUND STATION - LORA SERIAL RECEIVER BRIDGE [SIMULATION MODE]');
    console.log(`  Input Source  : STDIN (Piped / Interactive JSON lines)`);
    console.log(`  Target Backend: ${BACKEND_URL}`);
  } else {
    console.log('📡 SIH DRONE GROUND STATION - PHYSICAL LORA SERIAL RECEIVER BRIDGE (NODE.JS)');
    console.log(`  Serial Port  : ${SERIAL_PORT_NAME}`);
    console.log(`  Baud Rate    : ${BAUD_RATE}`);
    console.log(`  Target Backend: ${BACKEND_URL}`);
  }
}

main();

File: bridge/lora_bridge.py
#!/usr/bin/env python3
"""
=============================================================================
SIH Disaster-Response Drone System - Person 2 Ground Station
Physical LoRa Receiver Serial Bridge Script (Python)
=============================================================================
Architecture:
  Raspberry Pi (Drone) -> LoRa Transmitter RF -> Ground LoRa Receiver ->
  USB / Serial / UART -> THIS SCRIPT -> HTTP POST /api/events/lora -> Express Backend

Usage:
  python bridge/lora_bridge.py

Environment Variables:
  SERIAL_PORT : Serial port name (Default: 'COM3' on Windows, '/dev/ttyUSB0' on Linux)
  BAUD_RATE   : Baud rate (Default: 9600)
  BACKEND_URL : Express API endpoint (Default: 'http://localhost:5000/api/events/lora')
"""

import os
import sys
import json
import time
import requests

try:
    import serial
except ImportError:
    print("❌ ERROR: 'pyserial' library is not installed.")
    print("👉 Install it using: pip install pyserial requests")
    sys.exit(1)

# Environment configuration with defaults
SERIAL_PORT = os.getenv("SERIAL_PORT", "COM3" if os.name == "nt" else "/dev/ttyUSB0")
BAUD_RATE = int(os.getenv("BAUD_RATE", "9600"))
BACKEND_URL = os.getenv("BACKEND_URL", "http://localhost:5000/api/events/lora")

# Target backend connection timeout (seconds)
HTTP_TIMEOUT = 5

def validate_lora_packet(data):
    """
    Validate that the incoming JSON payload satisfies Person 1 LoRa API contract.
    """
    errors = []
    if not isinstance(data, dict):
        return False, ["Payload is not a valid JSON object"]

    if "hazard" not in data or data["hazard"] is None:
        errors.append("Missing required field 'hazard'")
    
    if "latitude" not in data or "longitude" not in data:
        errors.append("Missing required fields 'latitude' or 'longitude'")
    else:
        try:
            lat = float(data["latitude"])
            lng = float(data["longitude"])
            if not (-90 <= lat <= 90):
                errors.append(f"Latitude out of bounds: {lat}")
            if not (-180 <= lng <= 180):
                errors.append(f"Longitude out of bounds: {lng}")
        except (ValueError, TypeError):
            errors.append("Latitude and longitude must be valid floating point numbers")

    if errors:
        return False, errors
    return True, []

def process_serial_line(line_str):
    """
    Parse line string into JSON, validate fields, and post to backend API.
    """
    line_clean = line_str.strip()
    if not line_clean:
        return

    print(f"\n[PACKET RECEIVED] Raw input: {line_clean}")

    # Step 1: JSON Parsing
    try:
        packet = json.loads(line_clean)
        print(f"[PACKET PARSED] Event ID: {packet.get('event_id', 'AUTO')} | Hazard: {packet.get('hazard')} | Lat/Lng: ({packet.get('latitude')}, {packet.get('longitude')})")
    except Exception as e:
        print(f"❌ [MALFORMED PACKET] Failed to parse JSON: {e}")
        return

    # Step 2: Field Validation
    valid, validation_errors = validate_lora_packet(packet)
    if not valid:
        print(f"⚠️ [INVALID PAYLOAD] Packet failed validation rules: {', '.join(validation_errors)}")
        return

    # Step 3: Forward to Backend API
    try:
        response = requests.post(BACKEND_URL, json=packet, timeout=HTTP_TIMEOUT)
        if response.status_code in [200, 201]:
            print(f"✅ [PACKET FORWARDED] Backend HTTP {response.status_code}: {response.json().get('message')}")
        else:
            print(f"❌ [BACKEND ERROR] Backend HTTP {response.status_code}: {response.text}")
    except requests.exceptions.RequestException as req_err:
        print(f"❌ [BACKEND ERROR] Failed to reach backend endpoint ({BACKEND_URL}): {req_err}")

def main():
    print("=============================================================================")
    print("📡 SIH DRONE GROUND STATION - PHYSICAL LORA SERIAL RECEIVER BRIDGE")
    print("=============================================================================")
    print(f"  Role         : Person 2 Serial Hardware Bridge")
    print(f"  Serial Port  : {SERIAL_PORT}")
    print(f"  Baud Rate    : {BAUD_RATE}")
    print(f"  Target Backend: {BACKEND_URL}")
    print("=============================================================================\n")

    while True:
        try:
            print(f"🔌 Attempting serial connection to [{SERIAL_PORT}] at {BAUD_RATE} baud...")
            with serial.Serial(SERIAL_PORT, BAUD_RATE, timeout=2) as ser:
                print(f"✅ [RECEIVER CONNECTED] Successfully listening on {SERIAL_PORT}")
                print("   Ready for incoming LoRa packet frames from Person 1 drone...\n")

                while True:
                    line = ser.readline()
                    if line:
                        try:
                            decoded = line.decode('utf-8', errors='replace')
                            process_serial_line(decoded)
                        except Exception as decode_err:
                            print(f"⚠️ [MALFORMED PACKET] Read error: {decode_err}")

        except serial.SerialException as s_err:
            print(f"❌ [RECEIVER DISCONNECTED] Serial port error on {SERIAL_PORT}: {s_err}")
            print("🔄 Retrying connection in 5 seconds... (Press Ctrl+C to cancel)")
            time.sleep(5)
        except KeyboardInterrupt:
            print("\n🛑 Bridge script terminated by user.")
            sys.exit(0)

if __name__ == "__main__":
    main()
